clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java

```diff
@@ -71,12 +71,12 @@
     /**
      * @see KafkaConsumer#commit(Map, CommitType)
      */
-    public void commit(Map<TopicPartition, Long> offsets, CommitType commitType);
+    public void commit(Map<TopicPartition, OffsetMetadata> offsets, CommitType commitType);
```

Comment:
I feel enforcing users to create OffsetMetadata upon commit brings some overhead, especially in practice most people do not want to embed any metadata with their commits. I would like to propose an alternative solution regarding the API:

commit(Map<TopicPartition, Long>, ComitType, String /* Commit Message */)
commit(Map<TopicPartition, Long>, ComitType, String /* Commit Message */, ConsumerCommitCallBack)

And the commit message will be used as the metadata for all the partitions included. It is based on my assumption that users usually would have a single commit message (i.e. the metadata string) per each commit call; if they want to have different messages for different partitions (for example in CopyCat @ewencp), they can call commit() multiple times with the metadata strings on each partition. This of course may cause more round trips for all sync-commit, but I would suggest people to use the following pattern:

commit(async);
commit(async);
...
commit(sync);  // last call

Since now the commit calls are all ordered.